### ISSUE TITLE
The size of trained data with ConstantLengthDataset ?

### ISSUE THREAD

**Issue by AIchenkai** [lvwerra/trl#302]
This is a bit a trade-off between training efficiently and potentially corrupting a few examples: if you are limited by the number of examples rather than compute then you might want to avoid this and just have one text per sample in the batch and pad the batch. However, if you have a lot of data the above code is very efficient since there is no padding at all. True, some examples might not have the full context but usually in the long training this doesn't matter much. The strategy above is called packing and common practice when pretraining language models. Hope this helps!

_Originally posted by @lvwerra in https://github.com/lvwerra/trl/issues/291#issuecomment-1505045302_
         

Did you pack all dataset with ConstantLengthDataset in Supervised fine-tuning step? Or just using 5000 thousands examples as a "pre-training" method to add stack knowledge to LLaMA, then train most SE data in  rl_training?

**Comment by lvwerra** [comment]
We use packing only for the "pretraining" step and we have a dedicated split in the dataset for it: https://huggingface.co/datasets/lvwerra/stack-exchange-paired/tree/main/data

The dataset is quite big and we didn't use the full dataset in any of the steps.

**Comment by ToddMorrill** [comment]
@lvwerra, is there something within Hugging Face that prevents latter subsequences from attending to earlier subsequences when you use packing? If not, I'm a little surprised this works well enough. You might retain efficiency and high precision if you implemented attention masking so that subsequences only attend to tokens within the subsequence within a packed example. Interestingly, someone attempted a whole ICLR paper about this idea: https://openreview.net/forum?id=ZAzSf9pzCm

**Comment by liebkne** [comment]
I have the same question as @ToddMorrill 's.

**Comment by lvwerra** [comment]
The main reason is that there is nothing useful in previous sequences to attend to so the model probably learns very easily to not pay attention to previous sequences. Do you have any reason to believe otherwise?

**Comment by ToddMorrill** [comment]
I think these models will use any information available to them and to the extent that two sequences are related, the second sequence might attend to the first.

One experiment we can run to see if this is actually occurring is to look at the attention weights going from one example to another. But I think the more important question is to determine if something bad is occurring as a result of this, such as worse test-time performance or slower or less data-efficient learning. Showing this might involve training pairs of models (with and without packing) and measuring these properties.

**Comment by lvwerra** [comment]
> to the extent that two sequences are related

In general they are totally unrelated, no? so there is no information in a random sample from the dataset that would be useful for a second dataset. when in doubt you can always run with `packing=False` and have totally disconnected samples.

**Comment by ToddMorrill** [comment]
If they're two examples drawn from the same dataset it's definitely possible that they have some overlap in the task or subject matter being discussed. And packing=False is a great option if there's ever a real concern, but I'm just looking to have my cake and eat it too - namely to get the throughput gain of packing, while being certain that I'm not cross-contaminating. Thanks for weighing in, @lvwerra.

**Comment by lvwerra** [comment]
You could write your own logic that adds an attention mask matrix to masking everything before the EOS token :)
